core/src/main/scala/pureconfig/ConfigCursor.scala

```diff
@@ -57,7 +57,61 @@ sealed trait ConfigCursor {
    *         failures otherwise.
    */
   def asString: Either[ConfigReaderFailures, String] =
-    castOrFail(STRING, Set(NUMBER, BOOLEAN, STRING), { cv => String.valueOf(cv.unwrapped) })
+    castOrFail(STRING, v => Right(v.unwrapped.asInstanceOf[String]))
+
+  /**
+   * Casts this cursor to a boolean.
+   *
+   * @return a `Right` with the boolean value pointed to by this cursor if the cast can be done, `Left` with a list of
+   *         failures otherwise.
+   */
+  def asBoolean: Either[ConfigReaderFailures, Boolean] =
+    castOrFail(BOOLEAN, v => Right(v.unwrapped.asInstanceOf[Boolean]))
+
+  /**
+   * Casts this cursor to a long.
+   *
+   * @return a `Right` with the long value pointed to by this cursor if the cast can be done, `Left` with a list of
+   *         failures otherwise.
+   */
+  def asLong: Either[ConfigReaderFailures, Long] =
+    castOrFail(NUMBER, { v => (ConvertHelpers.catchReadError[Long](_.toLong) _)(v.unwrapped.toString) })
```

Comment:
For all numeric conversion methods, I'd rather see strategies that didn't involve converting the number to a string just to parse it again.

What do you think of relying on the internal knowledge that Typesafe Config only has `ConfigInt`, `ConfigLong` and `ConfigDouble` implementations and pattern-matching `v.unwrapped` on the primitives of those three cases instead? We can also try to do something with `Number` (since a `castOrFail(NUMBER, _)` is guaranteed to return a `ConfigNumber`), although it's not as clean.

Reply:
Makes sense. I've ended up using `Number` and playing along with its `*Value()` methods in 98499f5. This has the added benefit of allowing numbers represented in the decimal system but with no fractional part to be read as `Int`, `Long` or `Short`.